### PR TITLE
[4.4] Smart Search: Properly sort filters

### PR DESCRIPTION
**Pull request by Hackwar**
Pull Request for Issue #41930.

### Summary of Changes
The entries in the filters in Smart Search aren't properly ordered in the frontend.

### Testing Instructions
1. Create a website with some content, especially nested categories. Best solution would be to use the testing sampledata.
2. Go to the Smart Search search view in the frontend and open the advanced filters
3. Check for example the category filter

### Actual result BEFORE applying this Pull Request
The categories are pretty much random.

### Expected result AFTER applying this Pull Request
The categories are displayed properly like in the tree structure.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by ceford** [comment]
I have tested this item :white_check_mark: successfully on 55139c2bdc53459ab63c95d00c6559b20ec181a5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42835">issues.joomla.org/tracker/joomla-cms/42835</a>.</sub>

**Comment by richard67** [comment]
@Hackwar Shouldn’t this be fixed in the 4.4-dev branch? The issue was reported for 4.4.

**Comment by richard67** [comment]
> @richard67 See the last discussion we had about this. Nothing has changed since then.

@Hackwar So you think you as the JBS Team Lead and a member of the CMS Maintenance Team can do as you like and ignore all our policies?

**Comment by Hackwar** [comment]
I'm not ignoring policies. However I refuse to waste my time arguing if something is a bug and worthy of fixing in Joomla 4.4 and rather take the shortcut of providing that fix to 5.0+. Since the relevant maintainer decided to ignore this for over a week now as well, I'm assuming that it is fine the way I'm doing it.

**Comment by fgsw** [comment]
I have tested this item :white_check_mark: successfully on 55139c2bdc53459ab63c95d00c6559b20ec181a5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42835">issues.joomla.org/tracker/joomla-cms/42835</a>.</sub>

**Comment by richard67** [comment]
RTC as it has 2 successful tests. Release managers pls. decide if it should go into 5.0-dev or 4.4-dev.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42835">issues.joomla.org/tracker/joomla-cms/42835</a>.</sub>

**Comment by stimpsonjcat** [comment]
I've tested this on J5.1. I created some categories and subcategories in content and in contacts. I deliberately named some of the categories in contact and content categories the same.
At first it seemed to work. The sort order in the pull down menu is now correct. Is it intended that the duplicate category names from the two different components are the shown as one?

After some testing the search didn't work for content anymore. Even if I turn of all other search plugins, and leave content the only one turned on. I did rebuild the smart search index. The advanced search does show the content categories, but whatever search word I enter, there's no result. If I turn on search for contacts, the contacts will be found, but no articles will show up.

I revoked the patch, and still the content search doesn't work anymore.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42835">issues.joomla.org/tracker/joomla-cms/42835</a>.</sub>

**Comment by richard67** [comment]
Back to pending after rebase.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42835">issues.joomla.org/tracker/joomla-cms/42835</a>.</sub>

**Comment by adj9** [comment]
With J 5.1, the display after patching does not change.
I made a category and subcategory

##ACCESSIBILITY IDEA
I searched for the subcategory with its own prefix and found nothing, so I infer that the query does a search with the = operator. You can vary the search query with a 
LIKE '%''+<string>+'%'?

**Comment by Hackwar** [comment]
@stimpsonjcat Your issues are unrelated to this PR. This PR only touches the order of the entries in the filters.

@adj9 That is already an option in the component configuration of Smart Search. Also,if you only have 2 categories, there is nothing which could change, since they aren't out of order at that moment.

**Comment by stimpsonjcat** [comment]
I did test it again. It seemed that my articles had disappeared on the pizza-bugs-and-fun server.
I did test it again in 5.1 alpha 3. It works as expected. ✅

The only feature request I would suggest: if you have a top level category (i.e. red) and and a sublevel category (i.e. light red), I would expect, as an enduser, if I search in the top-level I would also find results in the sublevel.

**Comment by richard67** [comment]
> I did test it again. It seemed that my articles had disappeared on the pizza-bugs-and-fun server. I did test it again in 5.1 alpha 3 and retested on the testserver of today. It works as expected. ✅

@stimpsonjcat Could you use the blue "Test this" button in the issue tracker https://issues.joomla.org/tracker/joomla-cms/42835 to submit your test result? Otherwise it is not counted. Thanks in advance.

**Comment by stimpsonjcat** [comment]
I have tested this item :white_check_mark: successfully on 4a8e9a123474185f314b56b7cecb2c69b323bb9c<br /><br />Created multiple categories and subcategories in content and in contacts. 
Added some articles and contacts. Some contacts have the same word in the name as in the article. (I've written an article on Stefanie and added a contact named Stefanie)
In the frontend searched on Stefanie and clicked on the advanced search. The category dropdown now shows the categories in the correct alphabetical order and hierarchy. The categories with the same names in content and contacts are shown as one category. 


<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42835">issues.joomla.org/tracker/joomla-cms/42835</a>.</sub>

**Comment by eddiekonczal** [comment]
I have tested this item :white_check_mark: successfully on 4a8e9a123474185f314b56b7cecb2c69b323bb9c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42835">issues.joomla.org/tracker/joomla-cms/42835</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42835">issues.joomla.org/tracker/joomla-cms/42835</a>.</sub>

File: administrator/components/com_finder/src/Service/HTML/Filter.php [hunk 1]
```diff
@@ -487,4 +498,27 @@ public function dates($idxQuery, $options)
...
+    /**
+     * Method to flatten a tree to a sorted array
+     *
+     * @param   \stdClass[]  $array
+     *
+     * @return  \stdClass[]  Flat array of all nodes of a tree with the children after each parent
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    private function reduce($array)
```
Comment (laoneo) [inline administrator/components/com_finder/src/Service/HTML/Filter.php:511]: Can you typehint the $array attribute and give it a proper name like $nodes?

Comment (laoneo) [inline administrator/components/com_finder/src/Service/HTML/Filter.php]:
```suggestion
    private function reduce(array $array)
```

**Comment by laoneo** [comment]
Back to pending<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42835">issues.joomla.org/tracker/joomla-cms/42835</a>.</sub>

**Comment by ceford** [comment]
No good! With the patch applied I see the list in order but part of the Sample Data Articles is repeated.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42835">issues.joomla.org/tracker/joomla-cms/42835</a>.</sub>

**Comment by Hackwar** [comment]
Why do you want another test for this?

**Comment by ceford** [comment]
image

**Comment by richard67** [comment]
> I'm getting errors in Joomla 5.1.0-alpha4:

@stimpsonjcat How is this related to this PR here which is clearly made for 4.4-dev and not 5.1-dev?

**Comment by laoneo** [comment]
As we added only a type to a function argument, one test is enough. So merging it as it was RTC before.

**Comment by laoneo** [comment]
Thanks!

**Pull request merged**